Modules/overlapped.c

```diff
@@ -1338,6 +1408,184 @@ Overlapped_getpending(OverlappedObject *self)
+                         &handle, &bufobj, &flags, &AddressObj))
+        return NULL;
+
+    // Parse the "to" address
+    AddressLength = sizeof(AddressBuf);
+    AddressLength = parse_address(AddressObj, Address, AddressLength);
+    if(AddressLength < 0)
+        return NULL;
+
+    if(self->type != TYPE_NONE) {
+        PyErr_SetString(PyExc_ValueError, "operation already attempted");
+        return NULL;
+    }
+
+    if(!PyArg_Parse(bufobj, "y*", &self->user_buffer))
+        return NULL;
+
+#if SIZEOF_SIZE_T > SIZEOF_LONG
+    if(self->user_buffer.len > (Py_ssize_t)ULONG_MAX) {
+        PyBuffer_Release(&self->user_buffer);
+        PyErr_SetString(PyExc_ValueError, "buffer to large");
+        return NULL;
+    }
+#endif
+
+    self->type = TYPE_WRITE_TO;
+    self->handle = handle;
+    wsabuf.len = (DWORD)self->user_buffer.len;
+    wsabuf.buf = self->user_buffer.buf;
+
+    Py_BEGIN_ALLOW_THREADS
+        ret = WSASendTo((SOCKET)handle, &wsabuf, 1, &written, flags,
+                        Address, AddressLength, &self->overlapped, NULL);
+    Py_END_ALLOW_THREADS
+
+    self->error = err = (ret == SOCKET_ERROR ? WSAGetLastError() : ERROR_SUCCESS);
+
+    switch(err) {
+    case ERROR_SUCCESS:
+    case ERROR_IO_PENDING:
+        Py_RETURN_NONE;
+    default:
+        self->type = TYPE_NOT_STARTED;
+        return SetFromWindowsErr(err);
+    }
+}
+
+
+
+PyDoc_STRVAR(
+    Overlapped_WSARecvFrom_doc,
+    "RecvFile(handle, size, flags) -> Overlapped[(message, (host, port))]\n\n"
+    "Start overlapped receive");
+
+static PyObject *
+Overlapped_WSARecvFrom(OverlappedObject *self, PyObject *args) {
+    HANDLE handle;
+    DWORD size;
+    DWORD flags = 0;
+    DWORD nread;
+    PyObject *buf;
+    WSABUF wsabuf;
+    int ret;
+    DWORD err;
+
+    if(!PyArg_ParseTuple(args, F_HANDLE F_DWORD "|" F_DWORD,
+                         &handle, &size, &flags))
+        return NULL;
+
+    if(self->type != TYPE_NONE) {
+        PyErr_SetString(PyExc_ValueError, "operation already attempted");
+        return NULL;
+    }
+
+#if SIZEOF_SIZE_T <= SIZEOF_LONG
+    size = Py_MIN(size, (DWORD)PY_SSIZE_T_MAX);
+#endif
+    buf = PyBytes_FromStringAndSize(NULL, Py_MAX(size, 1));
+    if(buf == NULL)
+        return NULL;
+
+    wsabuf.len = size;
+    wsabuf.buf = PyBytes_AS_STRING(buf);
+
+    self->type = TYPE_READ_FROM;
+    self->handle = handle;
+    self->read_from.buffer = buf;
+    memset(&self->read_from.address, 0, sizeof(self->read_from.address));
+    self->read_from.address_length = sizeof(self->read_from.address);
+
+    Py_BEGIN_ALLOW_THREADS
+        ret = WSARecvFrom((SOCKET)handle, &wsabuf, 1, &nread, &flags,
+                          (SOCKADDR*)&self->read_from.address,
+                          &self->read_from.address_length,
+                          &self->overlapped, NULL);
+    Py_END_ALLOW_THREADS
+
+        self->error = err = (ret < 0 ? WSAGetLastError() : ERROR_SUCCESS);
```

Comment:
Some weird indentation going on around `Py_BEGIN_ALLOW_THREADS` and `Py_END_ALLOW_THREADS` in this file. (I know VS does strange things automatically, depending on what it thinks about the expansion, but normally we don't adjust indentation around these macros.)